### PR TITLE
24 implement b0 pbwo4 ecal

### PR DESCRIPTION
**Pull request by rahmans1**
### Briefly, what does this PR introduce?
Implement a PbWO4 calorimeter that can be used in the B0 region instead of the preshower.

### What kind of change does this PR introduce?
- [X] New feature (issue #24 )

### Please check if this PR fulfills the following:
- [X] Tests for the changes have been added
- [X] Documentation has been added / updated
- [X] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
User needs to modify [farforward.xml](https://github.com/eic/ip6/blob/24-implement-b0-pbwo4-ecal/ip6/far_forward.xml#L7) and replace B0_preshower.xml with B0_ECal.xml to run sims with the ECal

### Does this PR change default behavior?
We can decide later if we want to keep the preshower or the ECal as the default in the main geometry

### CODE REVIEW

File: ip6.xml [hunk 1]
```diff
@@ -43,6 +43,7 @@
     <constant name="ZDCEcal_ID"              value="163"/>
     <constant name="ZDCHcal_ID"              value="164"/>
     <constant name="VacuumMagnetElement_1_ID"        value="165"/>
+    <constant name="B0ECal_ID" value="166"/>
```
Comment (wdconinc) [inline ip6.xml:46]: This conflicts with what was introduced in #26. All changes to ID structures are best added as separate PRs. @shimasnd @rahmans1

File: ip6/far_forward.xml [hunk 2]
```diff
@@ -4,7 +4,10 @@
   <include ref="far_forward/ion_beamline.xml" />
   <include ref="far_forward/beampipe_hadron_B0.xml" />
   <include ref="far_forward/B0_tracker.xml"/>
-  <include ref="far_forward/B0_preshower.xml"/>
+  <!--
+         <include ref="far_forward/B0_preshower.xml"/>
+  -->
+  <include ref="far_forward/B0_ECal.xml"/>
```
Comment (wdconinc) [inline ip6/far_forward.xml:10]: Commented out code should be avoided.

Why comment it out?
- If it currently not working but will need to be used in the future:
  - file an issue with info on what is needed,
  - add the issue as a reference along with the commented out block.
- If is not going to be needed:
  - remove the code,
  - move the referred to `B0_preshower.xml` file into `unused/` in its directory.

File: src/B0ECal_geo.cpp [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+//////////////////////////////////////////////////
+
+using namespace std;
+using namespace dd4hep;
+
+static std::tuple<int, int> add_individuals(Detector& desc, Assembly& env, xml::Collection_t& plm,
+                                            SensitiveDetector& sens, int id);
+static std::tuple<int, int> add_disk(Detector& desc, Assembly& env, xml::Collection_t& plm, SensitiveDetector& sens,
+                                     int id);
+typedef ROOT::Math::XYPoint Point;
```
Comment (wdconinc) [inline src/B0ECal_geo.cpp:21]: Modernize: use `using`

File: src/B0ECal_geo.cpp [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+#include "DDRec/DetectorData.h"
+#include "DDRec/Surface.h"
+#include <XML/Helper.h>
+
+//////////////////////////////////////////////////
+// Far Forward B0 Electromagnetic Calorimeter
+//////////////////////////////////////////////////
+
+using namespace std;
+using namespace dd4hep;
```
Comment (wdconinc) [inline src/B0ECal_geo.cpp:15]: Error prone: avoid `using namespace std` and list only the entities you actually use frequently enough for this to make a difference in readability, e.g. `using std::vector`.

File: src/B0ECal_geo.cpp [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+static std::tuple<int, int> add_individuals(Detector& desc, Assembly& env, xml::Collection_t& plm,
+                                            SensitiveDetector& sens, int id);
+static std::tuple<int, int> add_disk(Detector& desc, Assembly& env, xml::Collection_t& plm, SensitiveDetector& sens,
+                                     int id);
```
Comment (wdconinc) [inline src/B0ECal_geo.cpp:20]: Preparation for merge geometry: Add these two geometry helpers to a separate utility header as in epic.

**Review by wdconinc — CHANGES_REQUESTED** [review]

**Comment by ajentsch** [comment]
How else should we toggle between the preshower and the EMCAL? We need both
geometries until a final decision is made by the project.

**Comment by wdconinc** [comment]
> How else should we toggle between the preshower and the EMCAL? We need both geometries until a final decision is made by the project.

Not disagreeing with the goal here, but this implementation leads to dead code in comment blocks. This is not the only area where we are making a decision between two detector types, but in other places we are not having to comment/uncomment includes.

**Comment by ajentsch** [comment]
Ahh okay, I see. What is the better way you'd prefer us to do it?

File: ip6.xml (hunk 1 above)
Comment (rahmans1) [reply ip6.xml:46]: For now, should I revert this commit? That should take away the conflict.  And introduce a separate PR for ID change?

**Comment by wdconinc** [comment]
Is one of the two major configurations for this production with preshower and one without? if so, I'd suggest using a `far_forward_arches.xml` and a `far_forward_brycecanyon.xml` for now.

**Comment by ajentsch** [comment]
Originally no, we weren't. But since Sakib got it done, I think we will. I
can make the configurations for the two options tomorrow morning - I also
need to figure out how to do things with the magnet settings. We will have
a bit of a challenge since some MC samples will literally be useless with
one magnet setting, and vice versa (meaning 18x275 GeV DVCS cannot be used
with 41 GeV magnet settings).

File: ip6.xml (hunk 1 above)
Comment (wdconinc) [reply ip6.xml:46]: You can just modify it to an ID that's not used yet even after the other changes.

File: ip6/far_forward.xml (hunk 2 above)
Comment (rahmans1) [reply ip6/far_forward.xml:10]: > Is one of the two major configurations for this production with preshower and one without? if so, I'd suggest using a `far_forward_arches.xml` and a `far_forward_brycecanyon.xml` for now.

Are you suggesting making two copies of far_forward.xml. And including the preshower in one and the emcal in the other? Doesn't epic by default look for the far_forward.xml?

Comment (wdconinc) [reply ip6/far_forward.xml:10]: Yes, but epic_brycecanyon could look for far_forward_brycecanyon etc

Comment (rahmans1) [reply ip6/far_forward.xml:10]: ok. understood. So, then there would need to be a separate PR to add epic_brycecanyon on the epic repo?

Comment (wdconinc) [reply ip6/far_forward.xml:10]: Yes. First prep here, probably add as matrix to tests for overlap, then file as PR to epic.

File: src/B0ECal_geo.cpp (hunk 5 above)
Comment (rahmans1) [reply src/B0ECal_geo.cpp]: Addressed by commit [57beb70](https://github.com/eic/ip6/pull/25/commits/57beb70dfec047f4d9491a1ae5f5908ed396e313), [4ccd587](https://github.com/eic/ip6/pull/25/commits/4ccd587ac78224d02aad1589446550577f99d132),  and [2a6d510](https://github.com/eic/ip6/pull/25/commits/2a6d510d2a9ce78cf8bda1dbb76207a8abff0b17). Please check if it's reasonable now.

File: ip6/far_forward.xml (hunk 2 above)
Comment (rahmans1) [reply ip6/far_forward.xml]: ok. i created the versioned far forward files [6f583d0](https://github.com/eic/ip6/pull/25/commits/6f583d07dd060d16778d90c349fff91f486abb71). Is there any thing else that needs to change within ip6 like the ip6.xml? I didn't understand what you meant by adding as matrix to tests for overlap.

File: src/B0ECal_geo.cpp (hunk 3 above)
Comment (rahmans1) [reply src/B0ECal_geo.cpp]: Addressed

**Review by wdconinc — APPROVED** [review]
This looks good.

Any idea why the required overlap checks are not reporting back?

**Comment by wdconinc** [comment]
> Any idea why the required overlap checks are not reporting back?

Ah, right, turning it into a matrix job changes the name so it doesn't match the requirements anymore. This should be all good then.

**Comment by wdconinc** [comment]
Updated branch with changes in master.

**Pull request merged**